Upgrade to support ES 5.x, refs #10847

jraddaoui opened the pull request:
Upgrade Elasticsearch support from 1.x to 5.x. Main changes:

- Update Elastica to v5.2.1 and add new required deps.: Psr-fig/log and
  elasticsearch-php. The later requires PHP >= v5.6.6, so that should be
  the min. PHP version for AtoM after this changes.
- Fix ES status check on install process.
- Fix ES error display (ElasticsearchException is deprecated).
- Update ES index mappings:
  - String fields can now be keyword (not analyzed) or text (analyzed) and
    the not analyzed/analyzed value is not needed in the index property.
  - Remove include_in_all inside multi fields.
  - Change usage of index_analyzer to analyzer.
  - Increase total fields limit to 2000.
- Update ES/Elastica queries:
  - Use setSize instead of setLimit.
  - Use setSource instead of setFields.
  - Filters are replaced by queries, which can be applied as filters using the
    setPostFilter method from the queries. There is no query for the missing
    and not filters, they have been replaced by using a must_not clause in a
    bool query.
  - Replace facets by aggregations: mostly the same with some simplifications
    in the populate method and the template partial. It removes the partially
    introduced hability of selecting multiple values. Some changes required in
    the unique records query for the language facet as the language query is
    no longer added as a filter. Drafts are now filtered in the query so there is
    no need to do it in the aggs.
- Fix empty resultset creation, a third param with the results is now needed.
- Fix search-box autocomplete display, the flat option is no longer needed
  in QubitHelper get_search_i18n.

I have not renamed the HTML classes and ids related to the facets to avoid
making a lot of changes in external customizations after the upgrade.

sbreker committed:
Update to Elastica 5.2.1, Refs# 10847

Updated to ruflin/Elastica 5.2.1.

sbreker committed:
Add Psr-fig/Log to vendor, Refs #10847

Elastica 5.1 is now dependent on Psr-fig/log. Adding this library to the
vendor directory. Psr-fig/Log uses the MIT License which is compatible with
AtoM.

sbreker committed:
Fix sfInstall issue ES 5 HTTP resp, Refs #10847

In Elasticsearch 5, checking status of ES by using a HTTP GET on the ES host
and port no longer returns the 'status' element. Get HTTP status code using
curl instead.

sbreker committed:
Update ES mapping, Refs #10847

This change includes updates to Elasticsearch mappings to account for changes
between ES 1.7 and 5.x.

Main changes:
- Change all type: string to 'text'
- Remove include_in_all inside multi fields
- Change usage of 'index_analyzer' to 'analyzer'

Additionally I have included a fix affecting i18n analyzers where a reference
was not being cleared - this resulted in incorrect language analyzers being
applied. (see 'unset' in arElasticSearchMapping.class.php)

sbreker committed:
Add elasticsearch-php 5.2.0 to vendor, Refs #10847

Elastica 5.2 is now dependent on elastic/elasticsearch-php. Adding this
library to the vendor directory. elasticsearch-php 5.2.0 is dual licensed
under Apache 2.0 and LGPL 2.1 which are compatible with AtoM.

sbreker committed:
Register namespace elasticsearch-php, Refs #10847

Add namespace registration for vendor/Elasticsearch API dependency.

jraddaoui committed:
setLimit -> setSize

jraddaoui committed:
Exists query instead of exists and missing filters

jraddaoui committed:
Stop using filters

Filters are replaced by queries, which can be applied as filters using the
setPostFilter method from the queries. There is no query for the missing and
not filters, they have been replaced by using a must_not clause in a bool query.

jraddaoui committed:
Fix ES error display

Elastica\Exception\ElasticsearchException is deprecated

jraddaoui committed:
Mapping changes

String fields can now be keyword (not analyzed) or text (analyzed) and
the not analyzed/analyzed value is not needed in the index property

https://www.elastic.co/blog/strings-are-dead-long-live-strings

jraddaoui committed:
Fix search-box autocomplete

Use setSource instead of setFields and fix display

jraddaoui committed:
Fix sorting

Use untouched field where needed and remove ignore_unmapped property
as we have an strict mapping and all the fields used to sort are declared

jraddaoui committed:
Remove unused facet functions from QubitApiAction

jraddaoui committed:
Disable facets

sevein commented:
Looking great so far! 👏

sevein commented on apps/qubit/modules/informationobject/templates/_cardViewResults.php:
```diff
@@ -11,15 +11,15 @@
 
       <a href="<?php echo url_for(array('module' => 'informationobject', 'slug' => $doc['slug'])) ?>">
-        <?php if (!empty($doc['digitalObject']['thumbnailPath'])
+        <?php if (isset($doc['digitalObject']) && !empty($doc['digitalObject']['thumbnailPath'])
```
Are you sure this change is necessary? empty() not enough? This seems to work for me:

```php
<?php

error_reporting(E_ALL);
ini_set('display_errors', '1');

$a = array();
empty($a['foo']['bar']);

$a = NULL;
empty($a['foo']['bar']);

jraddaoui replied:
If I leave it as it was I get the following notice, maybe it has something to be with how the sfOutputEscaperArrayDecorator works.:

```[Mon Jul 03 17:18:58.160718 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP Notice:  Undefined index: digitalObject in /home/radda/www/atom-2/vendor/symfony/lib/escaper/sfOutputEscaperArrayDecorator.class.php on line 119, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160724 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP Stack trace:, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160728 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP   1. {main}() /home/radda/www/atom-2/qubit_dev.php:0, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160731 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP   2. sfContext->dispatch() /home/radda/www/atom-2/qubit_dev.php:21, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160734 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP   3. sfFrontWebController->dispatch() /home/radda/www/atom-2/vendor/symfony/lib/util/sfContext.class.php:170, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160741 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP   4. sfController->forward() /home/radda/www/atom-2/vendor/symfony/lib/controller/sfFrontWebController.class.php:48, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160744 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP   5. sfFilterChain->execute() /home/radda/www/atom-2/vendor/symfony/lib/controller/sfController.class.php:238, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160747 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP   6. QubitTransactionFilter->execute() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfFilterChain.class.php:53, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160750 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP   7. sfFilterChain->execute() /home/radda/www/atom-2/lib/filter/QubitTransactionFilter.class.php:48, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160753 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP   8. sfHistoryPluginFilter->execute() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfFilterChain.class.php:53, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160756 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP   9. sfFilterChain->execute() /home/radda/www/atom-2/plugins/sfHistoryPlugin/lib/sfHistoryPluginFilter.class.php:18, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160759 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  10. sfRenderingFilter->execute() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfFilterChain.class.php:53, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160762 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  11. sfFilterChain->execute() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfRenderingFilter.class.php:33, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160765 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  12. QubitSettingsFilter->execute() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfFilterChain.class.php:53, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160767 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  13. sfFilterChain->execute() /home/radda/www/atom-2/lib/filter/QubitSettingsFilter.class.php:64, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160770 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  14. QubitLimitIpFilter->execute() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfFilterChain.class.php:53, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160772 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  15. sfFilterChain->execute() /home/radda/www/atom-2/lib/filter/QubitLimitIp.class.php:39, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160775 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  16. QubitSslRequirementFilter->execute() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfFilterChain.class.php:53, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160778 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  17. sfFilterChain->execute() /home/radda/www/atom-2/lib/filter/QubitSslRequirementFilter.class.php:35, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160780 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  18. QubitMeta->execute() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfFilterChain.class.php:53, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160785 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  19. sfFilterChain->execute() /home/radda/www/atom-2/lib/filter/QubitMeta.class.php:31, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160787 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  20. QubitLimitResults->execute() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfFilterChain.class.php:53, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160790 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  21. sfFilterChain->execute() /home/radda/www/atom-2/lib/filter/QubitLimitResults.class.php:34, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160793 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  22. sfExecutionFilter->execute() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfFilterChain.class.php:53, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160795 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  23. sfExecutionFilter->handleView() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfExecutionFilter.class.php:47, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160798 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  24. sfExecutionFilter->executeView() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfExecutionFilter.class.php:116, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160801 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  25. sfPHPView->render() /home/radda/www/atom-2/vendor/symfony/lib/filter/sfExecutionFilter.class.php:155, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160804 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  26. sfPHPView->renderFile() /home/radda/www/atom-2/vendor/symfony/lib/view/sfPHPView.class.php:185, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160806 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  27. require() /home/radda/www/atom-2/vendor/symfony/lib/view/sfPHPView.class.php:75, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160809 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  28. get_partial() /home/radda/www/atom-2/apps/qubit/modules/informationobject/templates/browseSuccess.php:338, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160812 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  29. sfPartialView->render() /home/radda/www/atom-2/vendor/symfony/lib/helper/PartialHelper.php:220, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160814 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  30. sfPHPView->renderFile() /home/radda/www/atom-2/vendor/symfony/lib/view/sfPartialView.class.php:124, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160817 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  31. require() /home/radda/www/atom-2/vendor/symfony/lib/view/sfPHPView.class.php:75, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
[Mon Jul 03 17:18:58.160820 2017] [:error] [pid 30542] [client 192.168.56.1:55952] PHP  32. sfOutputEscaperArrayDecorator->offsetGet() /home/radda/www/atom-2/vendor/symfony/lib/view/sfPHPView.class.php:14, referer: http://192.168.56.128:81/atom-2/qubit_dev.php/informationobject/browse
```

sevein replied:
Ugh, right, `sfOutputEscaperArrayDecorator`... that's sad but nothing we can do!

jraddaoui committed:
Fix clipboard clear query

jraddaoui committed:
Fix empty resultset creation

A third param with the results is now needed

sevein commented on apps/qubit/modules/user/actions/clipboardAction.class.php:
```diff
@@ -41,7 +41,7 @@ public function execute($request)
     if (!isset($allSlugs[$this->entityType]) || !count($allSlugs[$this->entityType]))
     {
-      $resultSet = new \Elastica\ResultSet(new Elastica\Response(null), new Elastica\Query);
+      $resultSet = new \Elastica\ResultSet(new Elastica\Response(null), new Elastica\Query, array());
```
😅

jraddaoui committed:
Add term aggregations

Mostly the same as facets with some simplifications in the populate method
and the template partial. It removes the partially introduced hability of
selecting multiple values. Some changes required in the unique records query
for the language facet as the language query is no longer added as a filter.
Drafts are now filter in the query so there is no need to do it in the aggs.

jraddaoui committed:
Replace query facet by filter aggregation

jraddaoui committed:
Fix languages agg in term index page

jraddaoui committed:
Rename facet to aggregation where possible

jraddaoui committed:
Remove unused facets code

jraddaoui committed:
Remove flat option from get_search_i18n

It's no longer being used

jraddaoui committed:
Add elasticsearch-php dep. to COPYRIGHT

MikeFE commented on apps/qubit/modules/search/templates/_aggregation.php:
```diff
@@ -0,0 +1,36 @@
+<?php if (!isset($aggs[$name]) || (!isset($filters[$name]) && count($aggs[$name]) < 2)) return ?>
```
What is this doing? Does return here just make it not render the partial?

jraddaoui replied:
Sorry Mike, I forgot about this. Yes, it works like in an action.

MikeFE commented on apps/qubit/modules/search/templates/_aggregation.php:
```diff
@@ -0,0 +1,36 @@
+<?php if (!isset($aggs[$name]) || (!isset($filters[$name]) && count($aggs[$name]) < 2)) return ?>
+<?php $openned = (isset($sf_request->$name) || (isset($open) && $open && 0 < count($aggs[$name]))) ?>
```
opened

jraddaoui replied:
Thanks! Will fix.

jraddaoui committed:
Fix new  reference

jraddaoui commented:
Merged in https://github.com/artefactual/atom/commit/eabbac520d765a4a903e01f0ef77bc068d71b7a7
